hub/core/meta/encode/base_encoder.py

```diff
@@ -0,0 +1,177 @@
+
+        Returns:
+            int: The index of the corresponding row inside the encoded state.
+        """
+
+        # TODO: optimize this (should accept an optional argument for starting point, instead of random binsearch)
+
+        if self.num_samples == 0:
+            raise IndexError(
+                f"Index {local_sample_index} is out of bounds for an empty byte position encoding."
+            )
+
+        if local_sample_index < 0:
+            local_sample_index += self.num_samples
+
+        return np.searchsorted(
+            self._encoded[:, self.last_index_index], local_sample_index
+        )
+
+    def register_samples(self, item: Any, num_samples: int):
+        """Register `num_samples` as `item`. Combines when the `self._combine_condition` returns True.
+        This method adds data to `self._encoded` without decoding.
+
+        Args:
+            item (Any): General input, will be passed along to subclass methods.
+            num_samples (int): Number of samples that have `item`'s value. Will be passed along to subclass methods.
+        """
+
+        # TODO: optimize this
+
+        self._validate_incoming_item(item, num_samples)
+
+        if self.num_samples != 0:
+            if self._combine_condition(item):
+                last_index = self._encoded[-1, self.last_index_index]
+                new_last_index = self._derive_next_last_index(last_index, num_samples)
+
+                self._encoded[-1, self.last_index_index] = new_last_index
+
+            else:
+                decomposable = self._make_decomposable(item)
+
+                last_index = self._encoded[-1, self.last_index_index]
+                next_last_index = self._derive_next_last_index(last_index, num_samples)
+
+                shape_entry = np.array(
+                    [[*decomposable, next_last_index]], dtype=ENCODING_DTYPE
+                )
+
+                self._encoded = np.concatenate([self._encoded, shape_entry], axis=0)
+
+        else:
+            decomposable = self._make_decomposable(item)
+            self._encoded = np.array(
+                [[*decomposable, num_samples - 1]], dtype=ENCODING_DTYPE
+            )
+
+    def _validate_incoming_item(self, item: Any, num_samples: int):
+        """Raises appropriate exceptions for when `item` or `num_samples` are invalid.
+        Subclasses should override this method when applicable.
+
+        Args:
+            item (Any): General input, will be passed along to subclass methods.
+            num_samples (int): Number of samples that have `item`'s value. Will be passed along to subclass methods.
+
+        Raises:
+            ValueError: For the general case, `num_samples` should be > 0.
+        """
+
+        if num_samples <= 0:
+            raise ValueError(f"`num_samples` should be > 0. Got: {num_samples}")
+
+    def _combine_condition(self, item: Any) -> bool:
+        """Should determine if `item` can be combined with the last row in `self._encoded`."""
+
```

Comment:
```suggestion
        """Should determine if `item` can be combined with the last row in `self._encoded`."""
        return False
```
Why no default for this one?

Reply:
because it should always be implemented specifically.

Reply:
the encoding is no longer an encoding without this condition

Reply:
I'm not entirely sure I understand, but in this limited scope I like "return False" better than the implicit "return None". Better to at least be consistent with the typing if one forgets to implement this override.

Reply:
it's not an implicit return none, it's an ABC class so it will raise not implemented